File: LintCode/480_binary-tree-paths/binary-tree-paths.java
/*
@Copyright:LintCode
@Author:   disillusion028
@Problem:  http://www.lintcode.com/problem/binary-tree-paths
@Language: Java
@Datetime: 16-11-23 07:47
*/

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root the root of the binary tree
     * @return all root-to-leaf paths
     */
    public List<String> binaryTreePaths(TreeNode root) {
        // Write your code here
            List<String> result = new ArrayList<String>();
            if(root==null) return result;
            String str = Integer.toString(root.val);
            findPaths(root,str,result);
            return result;
        }

        public void findPaths(TreeNode root, String str,List<String> result){
            String left= new String();
            String right = new String();
            if(root.left!=null)
            {
                left = str+"->"+Integer.toString(root.left.val);
                this.findPaths(root.left,left, result);
            }

            if(root.right!=null)
            {
                right = str+"->"+Integer.toString(root.right.val);
                this.findPaths(root.right,right,result);
            }
            if(root.right==null&&root.left==null) result.add(str);
        }
}

File: LintCode/480_binary-tree-paths/binary-tree-paths.cpp
/*
@Copyright:LintCode
@Author:   disillusion028
@Problem:  http://www.lintcode.com/problem/binary-tree-paths
@Language: C++
@Datetime: 16-11-23 07:59
*/

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root the root of the binary tree
     * @return all root-to-leaf paths
     */
    vector<string> binaryTreePaths(TreeNode* root) {
        // Write your code here
        
        vector<string> ret;
        if(root==NULL) return ret;
        
        string s="";
        Paths(root,ret,s);
        return ret;
    }
    
    void Paths(TreeNode *root,vector<string> &ret,string s){
        if(NULL==root) return;
        if(root->left==NULL && root->right==NULL){
            if(s=="") s+=to_string(root->val);
            else s+="->"+to_string(root->val);
            ret.push_back(s);
            s="";
            return ;
        }
        if(s=="") s+=to_string(root->val);
        else s+="->"+to_string(root->val);
        Paths(root->left,ret,s);
        Paths(root->right,ret,s);
    }
};